### PR TITLE
New function : Remove-DbaDbView

### PR DESCRIPTION
**Pull request by MikeyBronowski**
## Type of Change

 - [ ] Bug fix (non-breaking change, fixes # )
 - [x] New feature (non-breaking change, adds functionality, fixes # )
 - [ ] Breaking change (effects multiple commands or functionality, fixes # )
 - [x] Ran manual Pester test and has passed (`.\tests\manual.pester.ps1)
 - [ ] Adding code coverage to existing functionality
 - [x] Pester test is included
 - [ ] If new file reference added for test, has is been added to github.com/sqlcollaborative/appveyor-lab ?
 - [ ] Nunit test is included
 - [x] Documentation
 - [ ] Build system
 

### Purpose
 
Expanding functionality to manage database views.

### Commands to test

```powershell
Get-DbaDbView -SqlInstance localhost -Database Mikey -View Ocean | Remove-DbaDbView

or


Remove-DbaDbView -SqlInstance localhost -Database Mikey -View Ocean -Confirm:$false
```

### CODE REVIEW

File: functions/Remove-DbaDbView.ps1
Comment (wsmelton) [inline functions/Remove-DbaDbView.ps1:109]: Use of this pattern is not required in every command. I know we had a few that started using it from contributors but is not part of usual template of commands.

Use of this block also convolutes it more than it needs to be. PowerShell inherits the ability to bind parameter values based on both property name or value. It goes through this process regardless of what you have set on the parameter attributes.

In the end this command only needs to accept input from `Get-DbaDbView` and not a database, server or the DbaInstance parameter itself.

File: functions/Remove-DbaDbView.ps1
Comment (wsmelton) [inline functions/Remove-DbaDbView.ps1:93]:
```suggestion
        [Microsoft.SqlServer.Management.Smo.View[]]$InputObject,
```

This resolves any need to having to check what type is the input object. The object we output from `Get-DbaDbView` includes SqlInstance name and PowerShell will bind that for us without any hassle or extra code.

https://github.com/sqlcollaborative/dbatools/blob/de024aaa616fe37de7c8eedb796a2352ab08529a/functions/Get-DbaDbView.ps1#L167-L175

Comment (MikeyBronowski) [reply functions/Remove-DbaDbView.ps1:93]: Would that work just remove- standalone, without piping from get-

I would probably use get | remove myself as "safer" option anyway.

Comment (wsmelton) [reply functions/Remove-DbaDbView.ps1:93]: If you are calling remove function by itself you don't use the input object parameter.

File: functions/Remove-DbaDbView.ps1
Comment (MikeyBronowski) [reply functions/Remove-DbaDbView.ps1:109]: Makes sense. Do you know any "model" function(s) that follow the best practices and desired coding approach I could have a look at?

File: functions/Remove-DbaDbView.ps1
Comment (MikeyBronowski) [reply functions/Remove-DbaDbView.ps1]: Do my changes make more sense now?

File: functions/Remove-DbaDbView.ps1
Comment (andreasjordan) [inline functions/Remove-DbaDbView.ps1:106]: $InputObject is not a list of databases, but a list of views. So I would call the variable $view.

Comment (MikeyBronowski) [reply functions/Remove-DbaDbView.ps1:106]: Good catch

File: functions/Remove-DbaDbView.ps1
Comment (andreasjordan) [inline functions/Remove-DbaDbView.ps1:102]: I don't see any need in this foreach loop, as Get-DbaDbView can get the array in $SqlInstance as -SqlInstance. You could just test with `if ($SqlInstance) { ... }`. I know that this is widely used, but maybe we can streamline the code here and in other functions.

File: functions/Remove-DbaDbView.ps1
Comment (andreasjordan) [inline functions/Remove-DbaDbView.ps1:110]: Why don't you just use `$db.Drop()`? Testing for just the Name without Schema would maybe return and drop more views (if there are views with this name in other schemas).

**Comment by andreasjordan** [comment]
And have a look at #6885.

File: functions/Remove-DbaDbView.ps1
Comment (andreasjordan) [inline functions/Remove-DbaDbView.ps1:108]: As more than one SqlInstance is allowed, maybe also add the instance (.Parent.Parent.Name) in the message.

File: functions/Remove-DbaDbView.ps1
Comment (MikeyBronowski) [reply functions/Remove-DbaDbView.ps1]: > And have a look at #6885.

Thank you - I will have a look. Sounds like a great discussion and examples. Just need a fresh head and coffee :)

**Comment by MikeyBronowski** [comment]
> And have a look at #6885.

Thank you - I will have a look. Sounds like a great discussion and examples. Just need a fresh head and coffee :)

File: functions/Remove-DbaDbView.ps1
Comment (andreasjordan) [inline functions/Remove-DbaDbView.ps1:98]: But the parameter documentation says: "The database(s) to process. This list is auto-populated from the server. If unspecified, all databases will be processed." What do you want?

File: functions/Remove-DbaDbView.ps1
Comment (andreasjordan) [inline functions/Remove-DbaDbView.ps1:10]: No, no pipeline input for SqlInstance.

File: functions/Remove-DbaDbView.ps1
Comment (andreasjordan) [inline functions/Remove-DbaDbView.ps1:28]: You don't use this parameter at the moment.

File: functions/Remove-DbaDbView.ps1
Comment (MikeyBronowski) [reply functions/Remove-DbaDbView.ps1:108]: > As more than one SqlInstance is allowed, maybe also add the instance (.Parent.Parent.Name) in the message.

That's currently my output when I run :
```powershell
Get-dbadbview -SqlInstance $s1,$s2 -ExcludeSystemView -Database Mikey | Remove-DbaDbView -Verbose -Confirm:$false
```

```
VERBOSE: Performing the operation "Removing the view [dbo].[a] in the database Mikey" on target "sql1".
VERBOSE: Performing the operation "Removing the view [dbo].[a] in the database Mikey" on target "sql2".
```

File: functions/Remove-DbaDbView.ps1
Comment (andreasjordan) [inline functions/Remove-DbaDbView.ps1:32]: Only Get-DbaDbView, no more Get-DbaDatabase

File: functions/Remove-DbaDbView.ps1
Comment (andreasjordan) [inline functions/Remove-DbaDbView.ps1:57]: I would remove the " here.

File: functions/Remove-DbaDbView.ps1
Comment (andreasjordan) [inline functions/Remove-DbaDbView.ps1:71]: You only allow views SMOs to be piped in...

Comment (MikeyBronowski) [reply functions/Remove-DbaDbView.ps1:71]: I still can remove views like that
`Remove-DbaDbView -SqlInstance $s1 -Database Mikey -View b,c -Confirm:$false`

File: functions/Remove-DbaDbView.ps1
Comment (MikeyBronowski) [reply functions/Remove-DbaDbView.ps1:28]: Yeah, the same about "Schema" was there originally, but the Get-DbaDbView does not have it either. Wanted to finish the Remove- first so it is satisfactory and then add more options to the Get-DbaDbView as well.

**Comment by wsmelton** [comment]
Overall I guess we need to decide, here since there is no issue for this:

`Remove-DbaDbSchema` got approved with iterating and taking input object of a Database (not the schema) https://github.com/sqlcollaborative/dbatools/blob/development/functions/Remove-DbaDbSchema.ps1

Whereas for the master key removal it only processes inputs from the Get master key command: https://github.com/sqlcollaborative/dbatools/blob/development/functions/Remove-DbaDbSchema.ps1

If we consider how would we do this for say tables or even stored procedures (if we chose to create one for that)
We do not have a whole lot of Remove functions yet for the various objects in a databases so now is the time to decide. // @potatoqualitee  @jpomfret

File: functions/Remove-DbaDbView.ps1
Comment (wsmelton) [reply functions/Remove-DbaDbView.ps1:28]: There is no reason to use this as the assumption is the user has decided either from providing the database name or piping what databases will be included or excluded.

File: dbatools.psd1
Comment (wsmelton) [inline dbatools.psd1:547]: Please remove this from the PR, maintainer control this changes to the manifest and psm1 files.

**Review by wsmelton — CHANGES_REQUESTED** [review]

File: functions/Remove-DbaDbView.ps1
Comment (andreasjordan) [reply functions/Remove-DbaDbView.ps1:108]: Oh, yes - I forgot the target.

File: dbatools.psd1
Comment (MikeyBronowski) [reply dbatools.psd1:547]: Sorry, I must have clicked the wrong button :)

File: functions/Remove-DbaDbView.ps1
Comment (wsmelton) [inline functions/Remove-DbaDbView.ps1:22]: We can remove this function as it is not used in our other Remove functions but also they should control via input or Database the exact ones to use or not.

File: tests/Remove-DbaDbView.Tests.ps1
Comment (wsmelton) [inline tests/Remove-DbaDbView.Tests.ps1:22]:
```suggestion
        $null = New-DbaDatabase -SqlInstance $server -Name $dbname1
```

File: functions/Remove-DbaDbView.ps1
Comment (wsmelton) [reply functions/Remove-DbaDbView.ps1:10]: Yes, all of our commands should accept input from a string or `Get-Content serverlist.txt | Verb-Dba...`

This should not be removed and is the template description used across all of our public functions.

**Comment by wsmelton** [comment]
If you look at the process you'd use in an environment for removing objects from n+1 databases I would think it best to have the InputObject accept a Database class, but I can see cases where users want to select them based on the parent Get function of the object.

**Comment by andreasjordan** [comment]
I would only allow piped input of type of object to remove. This way we only have two parameter sets: Using the command with the same parameters as the Get- command (which will internally just call get Get- command with these parameters and then drop the returning objects) or looping through the piped objects and just drop them.

I personally think that a lot of the commands that allow different piped input objects and a mix of piped and parameter input are not deeply tested - especially with arrays as piped inputs and the additional use of parameters with arrays. On the long run I would suggest some breaking changes to reduce functionality but improve flawlessness. But this is my personal opinion and no one has to follow me.

**Comment by MikeyBronowski** [comment]
@wsmelton @andreasjordan Thank you for all the comments and reviews. I will park it for now until the "standard" is decided for Remove-* functions.
Do you want me to remove (oh irony) the Remove- function from #7313  until this decision is made?

**Comment by andreasjordan** [comment]
I would just wait with further work on #7313, maybe change it to a draft.

**Comment by MikeyBronowski** [comment]
> I would just wait with further work on #7313, maybe change it to a draft.

I think the only concern for that one was for the .Drop() method in the Remove-* function, the other two were ok?

**Comment by potatoqualitee** [comment]
I will need a better brain day to think about the Removes that we'll do, but I'm leaning pscustomobject. I do that with  my newer modules 👍🏼

**Comment by jpomfret** [comment]
Hey folks,
Had a think about this and I think I'd personally go for piping from `Get-` into the `Remove-` so you pass an object of the thing you want to get rid of into the remove. However, @potatoqualitee I'm interested in seeing an example of using PSCustomObject - I'm happy to be convinced otherwise if there is a better way.

I think my preference on that is because I'm thinking about DscResources and how the Get\Set\Test kind of fit together.

Side note - are these templates up-to-date for our current preferences and standards?  If not once we decide on how to handle this I'm happy to work on updating those so we can easily point people to them

https://github.com/sqlcollaborative/dbatools-templates/

**Comment by potatoqualitee** [comment]
Agreed, the piped input should be type of object to remove.

File: functions/Remove-DbaDbView.ps1
Comment (andreasjordan) [inline functions/Remove-DbaDbView.ps1:98]: This is not needed and dangerous. Just use the view object you already have: $vw.Drop()

File: functions/Remove-DbaDbView.ps1
Comment (andreasjordan) [inline functions/Remove-DbaDbView.ps1:90]: What about `$InputObject = Get-DbaDbView @PSBoundParameters`? This way only the used parameters are send, what prevents problems when the called function uses Test-Bound.

File: functions/Remove-DbaDbView.ps1
Comment (andreasjordan) [inline functions/Remove-DbaDbView.ps1:86]: I would like to have a begin block with a test that only $InputObject XOR $SqlInstance can be used.

**Comment by andreasjordan** [comment]
I think we are very close to finishing this.

File: functions/Remove-DbaDbView.ps1
Comment (andreasjordan) [reply functions/Remove-DbaDbView.ps1:86]: Ups. Not in the begin block, but at the beginning of the process block.

File: functions/Remove-DbaDbView.ps1
Comment (MikeyBronowski) [reply functions/Remove-DbaDbView.ps1:90]: Was not aware of that. I will give it a go next week :) Thank you

**Comment by MikeyBronowski** [comment]
> I think we are very close to finishing this.

Thanks for all the reviews @andreasjordan . It was purely based on some existing function, but I am learning here a lot

**Comment by andreasjordan** [comment]
Hi @MikeyBronowski , can I help you with this?

**Comment by potatoqualitee** [comment]
I'll ping @MikeyBronowski on Twitter 😊

**Comment by MikeyBronowski** [comment]
> Hi @MikeyBronowski , can I help you with this?

I will make some changes today and run tests. Appreciate your help!

**Comment by andreasjordan** [comment]
The last Remove-DbaDb* command we redesigned was Remove-DbaDbSequence: https://github.com/sqlcollaborative/dbatools/blob/development/functions/Remove-DbaDbSequence.ps1

We had some good discussions (#7408) about the functionallity and the style. And I would like to use this as a role model for similar commands. So can you please have a look and try to align this command with Remove-DbaDbSequence?

So basic principles are:
* Two ways of input: Pipelining from Get- command or using same parameters as Get- command. In the later case just issue Get- command to get the objects to remove.
* PSCustomObject as output for every object with properties Status and IsRemoved to indicate success or failure.

**Comment by MikeyBronowski** [comment]
> The last Remove-DbaDb* command we redesigned was Remove-DbaDbSequence: https://github.com/sqlcollaborative/dbatools/blob/development/functions/Remove-DbaDbSequence.ps1
> 
> We had some good discussions (#7408) about the functionallity and the style. And I would like to use this as a role model for similar commands. So can you please have a look and try to align this command with Remove-DbaDbSequence?
> 
> So basic principles are:
> 
> * Two ways of input: Pipelining from Get- command or using same parameters as Get- command. In the later case just issue Get- command to get the objects to remove.
> * PSCustomObject as output for every object with properties Status and IsRemoved to indicate success or failure.

oh, ok. I will have a look then.

File: functions/Remove-DbaDbView.ps1
Comment (wsmelton) [inline functions/Remove-DbaDbView.ps1:7]:
```suggestion
        Drops a database view(s), with supported piping from Get-DbaDbView.
```

File: functions/Remove-DbaDbView.ps1
Comment (wsmelton) [inline functions/Remove-DbaDbView.ps1:4]:
```suggestion
        Removes a database view(s) from each database and SQL Server instance.
```

File: functions/Remove-DbaDbView.ps1
Comment (andreasjordan) [inline functions/Remove-DbaDbView.ps1:10]: This text has to be changed, as the parameter does not accept pipeline input.

File: functions/Remove-DbaDbView.ps1
Comment (andreasjordan) [inline functions/Remove-DbaDbView.ps1:24]: The Get- command has a parameter ExcludeSystemView - what do you think: Would it be worth adding this here as well?

File: functions/Remove-DbaDbView.ps1
Comment (andreasjordan) [inline functions/Remove-DbaDbView.ps1:55]: Can you add a second example to show the non-piped usage?

File: tests/Remove-DbaDbView.Tests.ps1
Comment (andreasjordan) [inline tests/Remove-DbaDbView.Tests.ps1:20]: I think this line should be removed - as tests migth run in parallel.

File: tests/Remove-DbaDbView.Tests.ps1
Comment (andreasjordan) [inline tests/Remove-DbaDbView.Tests.ps1:19]: When we put the returnd Server SMO from Connect-DbaInstance into a variable - we call it $server.

File: tests/Remove-DbaDbView.Tests.ps1
Comment (andreasjordan) [inline tests/Remove-DbaDbView.Tests.ps1:31]: That works? Without the info in which instance the database should be removed?

Comment (MikeyBronowski) [reply tests/Remove-DbaDbView.Tests.ps1:31]: Good call.

File: tests/Remove-DbaDbView.Tests.ps1
Comment (potatoqualitee) [reply tests/Remove-DbaDbView.Tests.ps1:20]: This was added by me because it helps with appveyor's limited resources (tho the new experimental path may help with exhaustion). Tests do not run in parallel as they are isolated to synchronous commands in a VM. Please leave it in. 

Also, I'd like to prioritize the merging of this PR as it may cause burnout in the author and he used my commands as a baseline 🤗

File: functions/Remove-DbaDbView.ps1
Comment (potatoqualitee) [reply functions/Remove-DbaDbView.ps1:24]: Good idea! However, we can add it in a later PR, let's get this one merged.

Comment (andreasjordan) [reply functions/Remove-DbaDbView.ps1:24]: I agree.

File: tests/Remove-DbaDbView.Tests.ps1
Comment (potatoqualitee) [reply tests/Remove-DbaDbView.Tests.ps1:19]: using $instance2 is a common pattern in our tests. We can leave as is.

File: tests/Remove-DbaDbView.Tests.ps1
Comment (andreasjordan) [reply tests/Remove-DbaDbView.Tests.ps1:20]: Thanks for the info. So no change needed here.

File: tests/Remove-DbaDbView.Tests.ps1
Comment (potatoqualitee) [reply tests/Remove-DbaDbView.Tests.ps1:31]: Once this test is fix and the example is added, this will be mergeable. I'll take care of it!

File: tests/Remove-DbaDbView.Tests.ps1
Comment (andreasjordan) [reply tests/Remove-DbaDbView.Tests.ps1:19]: Ok.

File: functions/Remove-DbaDbView.ps1
Comment (potatoqualitee) [reply functions/Remove-DbaDbView.ps1:10]: I'll fix this as well.

Comment (potatoqualitee) [reply functions/Remove-DbaDbView.ps1:10]: We may have to fix this in other remove commands too in another PR

File: functions/Remove-DbaDbView.ps1
Comment (potatoqualitee) [reply functions/Remove-DbaDbView.ps1:60]: done! once tests pass, i should be able to merge unless you find any mistakes

**Review by andreasjordan — APPROVED** [review]
Yes!

**Comment by potatoqualitee** [comment]
Woo@ Merging and will publish shortly.

**Pull request merged**
